Add move and split redirects to the schema

ddbeck opened the pull request:
Towards #91 and the sequel to #2990.

~~Because this PR follows #2990, looking at the diff between them might be easier to look at. See: https://github.com/ddbeck/web-features/compare/91-jsonschema-source-of-truth...ddbeck:web-features:91-redirects~~

This PR does the following:

- Adds redirects to the JSON Schema
- Generates (and embellishes) types for the new schema
- Modifies our scripts to use the new types and support redirects
- Demonstrates both types of redirects, by splitting `single-color-gradients` and moving `numeric-seperators` to `numeric-separators` (https://github.com/web-platform-dx/web-features/pull/2484)

foolip commented on features/webcodecs.yml:
```diff
@@ -1,10 +1,5 @@
-name: WebCodecs
-description: The WebCodecs API provides low-level access to individual video frames and chunks of audio samples, for full control over the way media is processed.
-spec: https://w3c.github.io/webcodecs/
-caniuse: webcodecs
```
Let's use https://github.com/web-platform-dx/web-features/pull/2776 as the feature to demonstrating split. We'd split that into gradients (2 BCD keys) and conic-gradients (1 BCD key).

ddbeck replied:
I switched to the single color stop feature as our demonstrator in 9b276f04076f7afc187c0a847b0bdedd5f5fdee7

ddbeck commented:
@js

ddbeck commented:
@jcscottiii Would you mind taking this PR for a test drive, pointing your schema tooling at the JSON Schema file like you did in https://github.com/GoogleChrome/webstatus.dev/pull/1552 and https://github.com/web-platform-dx/web-features/pull/2990#issuecomment-2978084026? I'm hoping to get a move on with this schema change (and other requests, such as https://github.com/web-platform-dx/web-features/issues/1981).

This PR is stacked on top of #2990 and adds feature redirects (renames and splits). I'm wondering what the experience of consuming that version of the schema would be like.

<details>

<summary>spoilers</summary>

I found that, at least for TypeScript, Quicktype's generated types weren't as nice, so I wrote some convenience wrapper types in `types.ts`. I wonder if you'll be inclined to do the same thing.

</details>

jcscottiii commented:
Trying it out now!

jcscottiii reviewed (commented):
One question for you

ddbeck commented:
@jcscottiii I've added the `kind` the discriminator field, but the actual `discriminator` annotation appears to be an OpenAPI Spec extension rather than JSON Schema. I added a `$comment` field to document it (hopefully).

But, once I added the discriminator field, I found that the existing structure was duplicative. Using the discriminator meant I could flatten the feature data—this greatly simplifies distinguishing the types (even if Quicktype ignores this fact). We now have things like this:

```json
{ "kind": "feature", "name": "<abbr>", "description": "…" }
{ "kind": "moved", "redirect_target": "…" }
{ "kind": "split", "redirect_targets": ["feat1", "feat2"] }
```

I'm curious to know your thoughts on this variation.

ddbeck commented:
The actual schema validation is failing right now (but strangely, only for the split feature). Trying to figure out why exactly.

jcscottiii commented on schemas/data.schema.json:
```diff
@@ -273,117 +179,179 @@
+      },
+      "minItems": 1
+    },
+    "URLOrURLs": {
```
Nit: Currently this is the same as StringOrStrings. I think you meant to add the `format` of the string? Otherwise, could you simplify to use StringOrStrings.

jcscottiii reviewed (approved):
I have not reviewed everything else. But from the data.schema.json POV, it looks good to me! I only have one nit/question.

ddbeck commented:
> When I say partially split out, I mean Feature A still exists but it has been scoped down and now feature B and C are split out of Feature A.
> 
> I was writing some e2e tests for webstatus.dev and currently I don't see a way for Feature A to use both `"#/definitions/FeatureData"` and `"#/definitions/FeatureSplitData"`

@jcscottiii Good question! The main answer is: partial splits can't happen.

In some cases, we'll error correct a feature silently (e.g., if a few compat keys were assigned erroneously—that wouldn't impact the meaning of the name, description, mapping to caniuse, etc.—then we'll just move those keys to another feature). Otherwise, the feature will have to be fully split (i.e., the original ID is just a pointer to two or more others).

I've given some thoughts to more sophistication around partial splits, but it's something we'd have to consider in a v4.0.0 (along with a notion of aspects of a larger feature and/or composite features).

jcscottiii commented:
Sounds good to me!

Background: I was just thinking about adding a banner to the feature detail page on webstatus.dev to let users know that they could reference other features in the event the original feature's scope was downsized. But I couldn't see a way for the current schema to tell me that information directly.

In the meantime, if we really wanted this, we could potentially track the BCD keys. But that introduces its own challenges, such as determining whether a key's removal was an editorial correction or a true feature split.

Anyway, I'm completely okay leaving this for v4 once we have more usage of this.
